### ISSUE TITLE
fftn very inefficient for 2d fft of several images

### ISSUE THREAD

**Issue by zaccharieramzi** [numpy/numpy#14617]
I wanted to compute the fourier transform of several images. I was therefore benchmarking `numpy`'s `fft.fftn` against a brute force for loop. I noticed that when the "batch size" was the first dimension of my arrays, the brute force loop approach was faster. The bigger the batch size, the more significant the difference gets. I think that it suggests an inefficiency in the way `fftn` is computed. I had opened [an SO question](https://stackoverflow.com/questions/58009758/numpy-fftn-very-inefficient-for-2d-fft-of-several-images) on this matter but with an older `numpy` version, so the comments on [this answer](https://stackoverflow.com/a/58010962/4332585) are more relevant. 

### Reproducing code example:
A gist is available [here](https://gist.github.com/zaccharieramzi/f05930697b40b74b412e14b420316972), with batch size 1, to be able to reproduce this example in a Jupyter notebook more conveniently.

```python
import numpy as np

batch_size = 128
x_first = np.random.rand(batch_size, 256, 256)
x_last = np.random.rand(256, 256, batch_size)

def iterate_first_fft(arr):
    k = np.empty_like(arr, dtype=np.complex64)
    for i, a in enumerate(arr):
        k[i] = np.fft.fft2(a)
    return k

def iterate_last_fft(arr):
    k = np.empty_like(arr, dtype=np.complex64)
    for i in range(arr.shape[-1]):
        k[..., i] = np.fft.fft2(arr[..., i])
    return k

k_it_first = iterate_first_fft(x_first)
k_np_first = np.fft.fftn(x_first, axes=(1, 2))
np.testing.assert_allclose(k_it_first.real, k_np_first.real)
np.testing.assert_allclose(k_it_first.imag, k_np_first.imag)

k_it_last = iterate_last_fft(x_last)
k_np_last = np.fft.fftn(x_last, axes=(0, 1))
np.testing.assert_allclose(k_it_last.real, k_np_last.real)
np.testing.assert_allclose(k_it_last.imag, k_np_last.imag)
```

```python
%%timeit
k_it_first = iterate_first_fft(x_first)
```

```python
%%timeit
k_np_first = np.fft.fftn(x_first, axes=(1, 2))
```

```python
%%timeit
k_it_last = iterate_last_fft(x_last)
```

```python
%%timeit
k_np_last = np.fft.fftn(x_last, axes=(0, 1))
```

### Error message:

Output of the profiling:
```
# for iterate method with batch size first
199 ms ± 1.14 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
# for fftn method with batch size first
269 ms ± 7.38 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# for iterate method with batch size last
423 ms ± 12.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# for fftn method with batch size last
322 ms ± 6.41 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

### Numpy/Python version information:

```
1.17.2 3.6.8 (default, May  7 2019, 14:58:50) 
[GCC 5.4.0 20160609]
```

**Comment by mreineck** [comment]
I'm pretty sure that the overhead is caused by the high-level code in `numpy/fft/pocketfft.py`, which performs a lot of (partially unnecessary) rearranging when working on multi-D arrays. This is especially noticeable for relatively small FFT sizes like the ones in your test case. The slowdown should become less pronounced for bigger FFT lengths like 1024 etc.

I'm not sure how much time should be invested in tweaking this. In the longer term, `numpy` might consider using the FFT implementation from `scipy` 1.4, which is considerably faster and should not suffer from this rerranging overhead.

**Comment by zaccharieramzi** [comment]
So I tried with bigger images and indeed, the slowdown is not there anymore (`fftn` is more efficient than the for loop approach).
However, I am not entirely sure it has to do with `pocketfft` since, as I was pointing out in the [SO question](https://stackoverflow.com/questions/58009758/numpy-fftn-very-inefficient-for-2d-fft-of-several-images), with `numpy` version `1.16` the slowdown is even more present.

I tried with `scipy` version 1.4, and got the following results with the same batch size and image size as in [here](https://github.com/numpy/numpy/issues/14617#issue-500726528) (128, 256).

These are the results:
```
# for iterate method with batch size first
125 ms ± 3.1 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
# for fftn method with batch size first
140 ms ± 2.05 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
# for iterate method with batch size last
346 ms ± 4.98 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# for fftn method with batch size last
193 ms ± 9.86 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

The slowdown is indeed reduced (we still notice that batch size last is slower though). I'll let you close this.

**Comment by mreineck** [comment]
> However, I am not entirely sure it has to do with pocketfft since, as I was pointing out in the SO question, with numpy version 1.16 the slowdown is even more present.

I think we are actually trying to say the same thing :)
The contents of `pocketfft.py` in `numpy` 1.17 are actually very similar to `fftpack.py` in earlier releases; I just changed a few things where to adjust for `pocketfft` details, and I removed the plan caching. Most of the problematic array shuffling was not touched at all.

> The slowdown is indeed reduced (we still notice that batch size last is slower though). 

That's to be expected. With batch size last, the memory access patterns during the FFT operation are worse than for batch size first. This could be fixed by a full array transpose, but that in itself takes additional time (and space).

> I'll let you close this

I don't have the rights to do that; I'm only a contributor, not a member ;)

>  (if just before you could explain why you still think it's the multi-array re-arranging that's causing the slowdown).

I hope I managed to clarify that above. It must be some high-level memory shuffling since `pocketfft` itself (i.e. the C code used by `numpy`) only ever sees 1D transforms.
